Add ingress for test env

daleelR opened the pull request:
This PR is to add an inbound rule for the test env domain and also a new deploy pipeline

daleelR committed:
add ingress and deploy pipeline for test env.

daleelR committed:
make pipeline names unique

daleelR committed:
Use correct property name for values file.

daleelR committed:
temporarily remove new pipeline

daleelR committed:
revert pipeline name change

daleelR committed:
re=enable new pipeline.

daleelR committed:
make naming specific for test env.

daleelR committed:
Merge branch 'main' into tech/add-ingress-for-test-env

JWeston-HO commented on .drone.yml:
```diff
@@ -32,6 +34,44 @@ steps:
+      values_files:
+        - ./helm/test-values.yaml
+      release: callisto-ingress-nginx
+      kube_certificate: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUMyRENDQWNDZ0F3SUJBZ0lSQUxBdk1yVGVzQkhNUkliR0ZZa3JsaHN3RFFZSktvWklodmNOQVFFTEJRQXcKRlRFVE1CRUdBMVVFQXhNS2EzVmlaWEp1WlhSbGN6QWVGdzB4TnpBNE1qZ3hPVFU0TlRSYUZ3MHlOekE0TWpneApPVFU0TlRSYU1CVXhFekFSQmdOVkJBTVRDbXQxWW1WeWJtVjBaWE13Z2dFaU1BMEdDU3FHU0liM0RRRUJBUVVBCkE0SUJEd0F3Z2dFS0FvSUJBUURQNm5ZVGpmbmR2T2kxOEltYm5idk9aOXJ3c1hxSkZFTE1IVTkxT2V ...
```
I know we haven't for dev. But is it worth putting this certificate in a secret?

stephenkaye-ho replied:
Why are we putting it in a secret it's a public key

daleelR replied:
@JWeston-HO was this suggestion based around this ticket:
https://collaboration.homeoffice.gov.uk/jira/browse/EAHW-2321
which I think (ticket) was originally mentioned by @taomoh 

@stephenkaye-ho maybe we keep the public cert hardcoded here, and it can be reviewed as part of the above ticket?

stephenkaye-ho replied:
I wondered if this became a moot point because of a difference concern. Hardcoding it make it a bit more difficult should it ever need to change. Think of the repos we'd have to update. Therefore I was thinking a secret is good, not because it's sensitive but because it allows us to manage a very common setting more centrally. This is only true if we can add it at the "organisation" level though.

stephenkaye-ho replied:
@daleelR is going to look into adding the cert as an organisation secret so we can make it available across all pipelines and manage it in one place. Organisation secrets span across the HO so we think it will be useful to others on ACP but needs to be well named. (Example only acp-notprod-kube-api-certificate )

daleelR replied:
Neither myself or Stephen are able to add the secret to the organisation. A conversation with ACP to get it into there is ongoing. So for now the cert is being stored in the repo secrets.
If they agree to it, we can have a separate PR to just change the secret reference in here to the org one.

JWeston-HO replied:
Sounds good resolving conversation

laurieboyes commented on .drone.yml:
```diff
@@ -13,15 +13,17 @@ steps:
       chart: ./helm
-      values_file: ./helm/values.yaml
+      values_files:
+        - ./helm/values.yaml
```
Should we rename this file to `dev-values`? Maybe they want to be in a new `values/…` directory as well so they’re not spread out all over the project root 🤔

daleelR committed:
update default values file name to be specific to dev.
move value files into new dir.
use kube certificate from secret.
use not prod api url secret on dev pipeline.

stephenkaye-ho commented on .drone.yml:
```diff
@@ -13,25 +13,67 @@ steps:
+  arch: amd64
+
+steps:
+  - name: lint helm files
```
If we are promoting deployments linting shouldn't happen in the deploy step as we'd have to potentially go back into a branch to fix any identified  issues

stephenkaye-ho replied:
I understand now that the pipelines are always running so the lint step of each pipeline always runs, even in a branch. As long as this isn't impacted when the depends_on is added to make the deployments sequential that we don't have the issue that I was concerned with above

stephenkaye-ho commented on .drone.yml:
```diff
@@ -13,25 +13,67 @@ steps:
+      kube_token:
+        from_secret: test_kube_token
+    when:
+      event: push
```
Are we not promoting deployments?

daleelR replied:
I thought we were only promoting for the actual services where we're packaging builds.. 
Given these are more like 'one time'  env specific value changes for ingress I thought we'd want it to deploy on main.
Perhaps i misunderstood, so should we be in fact promoting these deployments?

laurieboyes replied:
Hey Stephen, it hadn’t actually occurred to us to have manual promotion gates in our ingress config, but I can definitely see the value in having a mechanism for ensuring ingress changes look ok in our not-prod environments before applying them to prod

Daleel and I reckon it would be overkill to have promotion steps between _all_ of our environments. We propose that for all environments other than prod, the changes are applied immediately after merge. We don’t want to be in a situation where we have to do, say, 4 different manual promotions for different environments every time we make a change (including rolling back).

stephenkaye-ho replied:
I was thinking less about the manual promotion that we have built into the service and more about the release process.
Ideally the promote to test should depend on the deploy to dev pipeline. If something was to be broken in a change, currently we would break both env. I would advocate we deploy to dev and when the pipeline successfully completes it will deploy to test. If deploying to dev fails, it should not deploy to test. I'd probably pee myself in excitement if the deployment could actually validate itself as well i.e. curl the ingress point to make sure we didn't break anything but I may have got ahead of myself

laurieboyes replied:
> I'd probably pee myself in excitement

😂 wonderful

That makes a lot of sense 👍  do you agree that a manual promote step before going to prod is a sensible idea while our automated validation isn’t yet something to get dangerously excited about?

daleelR replied:
Agreed, that makes sense. @stephenkaye-ho Regarding curling the ingress point for verification could we potentially add a new ticket to cover that which could maybe be pickup during some random downtime along the journey of this ticket and just depend on the dev pipeline success for now?

stephenkaye-ho replied:
I'm happy that we add verification as a new ticket. Regarding whether or not we have a manual promote step, I'd look back to why we added one to start with. I hope I get this right. The manual promote step was there to allow for QA to validate the work before it was deployed i.e. run automated or manual tests. Would they do this with the ingress configuration? Maybe one we would feel more comfortable ourselves having a manual step to allow us to verify the deployment before going to prod.
For me deploying sequentially through the environments increased my confidence and is easy to implement.

stephenkaye-ho replied:
Following conversation with @daleelR @holly-blues. We agreed that the following would be new tickets and therefore out of the scope of this PR.
1. Verifying deployments i.e. curl the endpoint
2. Manual promotion to production.

This PR will be updated to make the deployments for dev and test sequential

daleelR committed:
add (dead) ui and accruals test values

daleelR committed:
add depends on dev deploy to test deploy step

daleelR committed:
move depends on to pipeline level

daleelR committed:
use correct pipelien name

daleelR committed:
move depends on back to test deploy step

daleelR committed:
move depends on back to pipeline level.
Partially comment out dev values to purposely fail dev pipeline for verification.

daleelR committed:
un-comment dev values to pass dev pipeline.
update kube cert name.
